### ISSUE TITLE
Some of Impact's GUI changes will break Forge mods

### ISSUE THREAD

**Issue by LeafHacker** [ImpactDevelopment/ImpactClient#590]
Assuming Impact ever works with forge, it would be useful for other mods to be able to add buttons, etc to GUIs like the main menu. With the current implementation of `OverrideGuiMainMenu` this wouldn't be possible.

Where possible we should have less invasive GUI changes, injecting into the existing class or at least verifying that others' injections will work with our override classes

The main menu will be a bit more complex, assuming we wish to keep the custom look. It may be possible to get a list of vanilla buttons, then filter through them to find custom 3rd party buttons that we could display somewhere sensible.

We should also try to avoid breaking other stuff like liteloader's weird custom menu overlay thing with the popout button.

- [x] Custom Main Menu handle Forge
- [x] ~~Trick~~ Liteloader into hooking our Custom Main Menu
- [ ] Option to disable Custom Main Menu (#109) (partially complete)
- [x] ~~Events and hooks for injecting into GUI methods~~
- [x] ~~System for injecting into GUIs~~ _Just write a mixin for each GUI instead_
- [x] Don't override `GuiIngameMenu`
- [x] Don't override `GuiChat`
- [x] Don't override `GuiChest`
- [x] Don't override `GuiCustomizeSkin`
- [x] Don't override `GuiDisconnect`
- [x] Don't override `GuiGameOver`
- [x] Don't override `GuiMainMenu (custom option will still short-circuit methods)

/cc @babbaj

**Comment by babbaj** [comment]
yes very cool

also every mixin should be able to work with other mods

**Comment by LeafHacker** [comment]
@ZeroMemes has added a special case for Forge in the Main Menu which adds the forge info text in the bottom left, clicking this will open the mods menu.
![image](https://user-images.githubusercontent.com/23565638/40583052-02ab8206-617e-11e8-970b-c54999b21010.png)

We plan to make the custom main menu optional, so users can choose to use the vanilla menu with an injected Impact button instead.

I've made an initial stab at the basic infrastructure for hooking into GUIs in a similar style to our existing overrides, which should make the transition smoother.

Updated issue with a checklist.

**Comment by LeafHacker** [comment]
Custom Liteloader handling shouldn't be required since they [hook into](http://develop.liteloader.com/liteloader/LiteLoader/blob/655d8a93108ea48c11e2d73ad4ebc5350ece6ee2/src/client/java/com/mumfrey/liteloader/client/LiteLoaderPanelManager.java#L327) anything that is an instanceof `GuiMainMenu`:

```java
private boolean isPanelSupportedOnScreen(GuiScreen guiScreen)
{
    return (guiScreen instanceof GuiMainMenu || guiScreen instanceof GuiIngameMenu || guiScreen instanceof GuiOptions);
}
```

**Comment by LeafHacker** [comment]
Changes implemented in a 4.3 branch, pending polish and merge
